syntax = "proto3";

package robot.api.gateway.v1;

service RobotApiGateway {
    //Login
    rpc Login(LoginRequest) returns (LoginResponse);

    //Heartbeat
    rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);

    //status
    rpc Status(StatusRequest) returns (StatusResponse);

    //status
    rpc Pos(PosRequest) returns (PosResponse);

    //map upload (occupancy grid)
    rpc MapUpload(MapUploadRequest) returns (MapUploadResponse);

    //control
    //rpc RobotControl(stream RobotControlRequest) returns (RobotControlResponse);

    //webrtc
    //rpc HandleRTC(stream RtcRequest) returns (RtcResponse);
}

message LoginRequest {
    string robot_id=1;
    string robot_secret=2;
}

message LoginResponse {
    bool success = 1;
    string access_token=2;
    string refresh_token=3;
}

message HeartbeatRequest {
    string robot_id=1;
}

message HeartbeatResponse {
    bool success = 1;
    string result=2;
}

message StatusRequest {
    string robot_id=1;
    float battery=2;
    string status=3;
    string error=4;
}

message StatusResponse {
    bool success = 1;
    string result=2;
}

message PosRequest {
    string robot_id=1;
    double x=2;
    double y=3;
    double z=4;
    double orig_x=5;
    double orig_y=6;
    double orig_z=7;
    double orig_w=8;
    double linear_speed=9;
    double angular_speed=10;
}

message PosResponse {
    bool success = 1;
    string result=2;
}

message MapUploadRequest {
    string robot_id = 1;
    uint32 width = 2;
    uint32 height = 3;
    float resolution = 4;
    double origin_x = 5;
    double origin_y = 6;
    double origin_z = 7;
    double origin_w = 8;
    bytes data = 9;
    int64 timestamp_ms = 10;
    string frame_id = 11;
}

message MapUploadResponse {
    bool success = 1;
    string result = 2;
}
